designs/intrinsics_design.md

```diff
@@ -0,0 +1,391 @@
+
+This is resource specific and seperates regions like US and China into subgroups. This will be found in a large dictionary.
+
+### AWS::REGION
+
+This property will first be attempted to be read from the enviroment settings. If it’s not there, it becomes randomly chosen from the list of regions.
+
+### AWS::StackId
+
+Currently, there is no real way to figure this one out. It will be a randomly generated String
+
+### AWS::StackName
+
+Currently, there is no real way to figure this one out. It will be a randomly generated String
+
+### AWS::URLSuffix
+
+This will be replaces by the corresponding [amazonaws.com](http://amazonaws.com/) or [amazonaws.com.cn](http://amazonaws.com.cn/) depending on the region settings. This will be found in a dictionary. 
+
+## Implementing the Code
+
+This can be seperated into a Intrinsics resolver and a SymbolResolver.
+The IntrinsicsResolver will recursively parse the template and all the attributes and the symbol table will be injected in at runtime. 
+
+### Intrinsic Symbol Table
+
+* logical_id_translator: This will be used as an exact translation for translating pseudo/logical_id types
+
+* default_type_resolver: This will be in the following form, resolving the type
+
+```
+{
+    "AWS::ApiGateway::RestApi": {
+        "RootResourceId": "/"
+    }
+}
+```
+
+* common_attribute_resolver: resolves common attributes that will be true across all
+
+```
+{    "Ref": lambda p,r: "",
+     "Arn:": arn_resolver}
+```
+
+
+First pseudo types are checked. If item is present in the logical_id_translator it is returned.Otherwise, it falls back to the default_pseudo_resolverThen the default_type_resolver is checked, which has common attributes and functions for each types.Then the common_attribute_resolver is run, which has functions that are common for each attribute.
+
+### Intrinsic Resolver
+
+The will be the core part of the resolver, which will be recursively called.
+
+```
+def intrinsic_property_resolver():
+    if key in self.intrinsic_key_function_map:
+        # process intrinsic function
+    elif key in self.conditional_key_function_map:
+        # process conditional
+    else:
+        # In this case, it is a dictionary that doesn't directly contain an intrinsic resolver and must be
+        # re-parsed to resolve.
+```
+
+The template will be resolved item by item
+
+```
+processed_template = {}
+for key, val in self.resources.items():
+    processed_key = self.symbol_resolver.get_translation(key, IntrinsicResolver.REF) or key
+
+    processed_resource = self.intrinsic_property_resolver(val)
+    processed_template[processed_key] = processed_resource
+```
+
+If there is an error with the resource, the item will be ignored and processed.
+
+## Integration into SAM-CLI
+
+This can be very easily plugged into the current SAM-CLI code. This needs to be run after the SamBaseTranslator runs in the ApiProvider. This will go through every property and check if it requires intrinsic resolution. 
```

Comment:
`SamBaseTranslator` runs in any provider not just the ApiProvider.

Reply:
I was talking about how it can be integrated into the current codebase where this can be added into the ApiProvider